### PR TITLE
abci: change client to use multi-reader mutexes

### PR DESCRIPTION
**Pull request by tychoish**

### CODE REVIEW

File: abci/client/grpc_client.go [hunk 1]
```diff
@@ -68,8 +68,8 @@ func (cli *grpcClient) OnStart() error {
 	go func() {
 		// Use a separate function to use defer for mutex unlocks (this handles panics)
 		callCb := func(reqres *ReqRes) {
-			cli.mtx.Lock()
```
Comment (alexanderbez) [inline abci/client/grpc_client.go:71]: This should remain a write-lock, no?

**Review by alexanderbez — COMMENTED** [review]
I think we only want to make `Info` and `Query` calls read-locked. Everything else should stay write-locked if I'm not mistaken?

File: abci/client/grpc_client.go (hunk 1 above)
Comment (tychoish) [reply abci/client/grpc_client.go:71]: I can't see any data in the client it's writing/modifying.

**Comment by tychoish** [comment]
> Everything else should stay write-locked if I'm not mistaken?

Is there a reason for leaving calls that don't modify the client structs under write-locks?

File: abci/client/grpc_client.go (hunk 1 above)
Comment (alexanderbez) [reply abci/client/grpc_client.go:71]: True, but I just don't know enough about the ABCI clients to be comfortable giving this change an ACK or not :-/

File: abci/client/local_client.go [hunk 2]
```diff
@@ -76,8 +78,8 @@ func (app *localClient) InfoAsync(ctx context.Context, req types.RequestInfo) (*
 }
 
 func (app *localClient) DeliverTxAsync(ctx context.Context, params types.RequestDeliverTx) (*ReqRes, error) {
-	app.mtx.Lock()
-	defer app.mtx.Unlock()
+	app.mtx.RLock()
```
Comment (alexanderbez) [inline abci/client/local_client.go:81]: this must be a write-lock.

**Review by alexanderbez — CHANGES_REQUESTED** [review]

File: abci/client/grpc_client.go (hunk 1 above)
Comment (tychoish) [reply abci/client/grpc_client.go]: cool, I unwound a bunch

File: abci/client/local_client.go (hunk 2 above)
Comment (tychoish) [reply abci/client/local_client.go]: does this mean that application implementations are depending on being called within a lock that they don't manage? 

In any case, I've made these all use the write lock.

**Review by alexanderbez — APPROVED** [review]

**Comment by marbar3778** [comment]
@mergifyio backport v0.34.x

**Pull request merged**